src/rgw/rgw_data_sync.cc

```diff
@@ -1311,59 +1312,78 @@ class RGWDataSyncSingleEntryCR : public RGWCoroutine {
 
   int operate() override {
     reenter(this) {
-      do {
-        if (marker_tracker) {
-          marker_tracker->reset_need_retry(obligation.key);
+      if (state->obligation) {
+        // this is already syncing in another DataSyncSingleEntryCR
+        if (state->obligation->timestamp < obligation.timestamp) {
+          // cancel existing obligation and overwrite it
+          tn->log(10, SSTR("canceling existing obligation " << *state->obligation));
+          complete = std::move(*state->obligation);
+          *state->obligation = std::move(obligation);
+        } else {
+          // cancel new obligation
+          tn->log(10, SSTR("canceling new obligation " << obligation));
+          complete = std::move(obligation);
         }
-        tn->log(0, SSTR("triggering sync of source bucket/shard " << bucket_shard_str{state->key}));
+      } else {
+        // start syncing a new obligation
+        state->obligation = obligation;
 
-        yield call(new RGWRunBucketSourcesSyncCR(sc,
-                                                 std::nullopt, /* target_bs */
-                                                 state->key,
-                                                 tn, &progress));
-        if (retcode == 0) {
-          tn->log(20, SSTR("RunBucketSources progress=" << progress));
-        }
-      } while (marker_tracker && marker_tracker->need_retry(obligation.key));
+        do {
+          yield {
+            tn->log(4, SSTR("triggering sync of source bucket/shard " << bucket_shard_str{state->key} << " " << *state->obligation));
+
+            call(new RGWRunBucketSourcesSyncCR(sc,
+                                               std::nullopt, /* target_bs */
+                                               state->key,
+                                               tn, &progress));
+          }
+        } while (marker_tracker && marker_tracker->need_retry(obligation.key));
 
+        tn->log(10, SSTR("sync finished on " << bucket_shard_str{state->key}));
```

Comment:
Racy? Note that at this point here state->obligation() is not empty, but bucket sync could have already completed. That means that another call with the same source bucket would not trigger sync. Is that possible or ok because same source will always be handled by the same data sync shard?

Reply:
it's only racy if we yield this coroutine before the call to `state->obligation.reset()` just below

Reply:
i updated the commit to move this block directly under the while-loop:
```c++
        }
        // any new obligations will process themselves
        complete = std::move(*state->obligation);
        state->obligation.reset();
```